librocksdb_sys/crocksdb/c.cc

```diff
@@ -3779,4 +3800,110 @@ int crocksdb_keyversions_type(const crocksdb_keyversions_t *kvs, int index) {
   return kvs->rep[index].type;
 }
 
+struct ExternalSstFileModifier {
+  ExternalSstFileModifier(Env *env, ColumnFamilyData *cfd, DBOptions &db_options)
+  :env_(env), cfd_(cfd), env_options_(db_options), table_reader_(nullptr) { }
+
+  Status Open(std::string file) {
+    file_ = file;
+    // Get External Sst File Size
+    uint64_t file_size;
+    auto status = env_->GetFileSize(file_, &file_size);
+    if (!status.ok()) {
+      return status;
+    }
+
+    // Open External Sst File
+    std::unique_ptr<RandomAccessFile> sst_file;
+    std::unique_ptr<RandomAccessFileReader> sst_file_reader;
+    status = env_->NewRandomAccessFile(file_, &sst_file, env_options_);
+    if (!status.ok()) {
+      return status;
+    }
+    sst_file_reader.reset(new RandomAccessFileReader(std::move(sst_file), file_));
+
+    // Get Table Reader
+    status = cfd_->ioptions()->table_factory->NewTableReader(
+        TableReaderOptions(*cfd_->ioptions(), env_options_,
+                           cfd_->internal_comparator()),
+        std::move(sst_file_reader), file_size, &table_reader_);
+    return status;
+  }
+
+  Status SetGlobalSeqNo(uint64_t seq_no) {
+    if (table_reader_ == nullptr) {
+      return Status::InvalidArgument("File is not open or seq-no has been modified");
+    }
+    // Get the external file properties
+    auto props = table_reader_->GetTableProperties();
+    const auto& uprops = props->user_collected_properties;
+    // Validate version and seqno offset
+    auto version_iter = uprops.find(ExternalSstFilePropertyNames::kVersion);
+    if (version_iter == uprops.end()) {
+      return Status::Corruption("External file version not found");
+    }
+    uint32_t version = DecodeFixed32(version_iter->second.c_str());
+    if (version != 2) {
+      return Status::NotSupported("External file version should be 2");
+    }
+
+    auto seqno_iter = uprops.find(ExternalSstFilePropertyNames::kGlobalSeqno);
+    uint64_t ori_seq_no = DecodeFixed64(seqno_iter->second.c_str());
```

Comment:
Is it possible that DecodeFixed64 failed, or do we need to check the length of seqno_iter->second?

Reply:
DecodeFixed64 will not fail.
I think we don't because there is no check in `ExternalSstFileIngestionJob` either.